### PR TITLE
Stop font properties from leaking

### LINKED ISSUE #359: While using Unicode fonts if markdown in multi_cell is on, then the rendering of muliti_cell in a subsequent line is broken
This code renders a "broken" table if Unicode fonts are used:
```Python
    data = (
            ("First name", "Last name", "Age", "City"),
            ("Jules", "Smith", "34", "San Juan"),
               )
    l_height = pdf.font_size * 1.2
    cell_width = pdf.epw / 4

    pdf.set_fill_color(240)

    for j, row in enumerate(data):
        for cell in row:
            if j == 0:
                pdf.multi_cell(cell_width, l_height, '**' + cell + '**', border=1, fill=True, align='L', ln=3, max_line_height=l_height, markdown=True)
            else:
                pdf.multi_cell(cell_width, l_height, cell, border=1, align='L', ln=3, max_line_height=l_height)
        pdf.ln()
```

If you set "markdown=False" in the code above, the rendering is not broken any more.

This bug emerged after some latest changes in the master branch.

I am using Python 3.9.2 with the latest version of the master fpdf2 branch (2.5.1)

### PR DESCRIPTION
**Pull request by gmischler**
Fixes #359  (and possibly others)

**Checklist**:

- [x] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged 

- [na] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

- [x] The PR description or comment contains a picture of a cute animal (not mandatory but encouraged :frog:)

This solves the problems with leaking state after certain combinations of text output. It replaces several previous attempts to solve similar problems, and does so in a more straightforward and complete way.

* Using qQ when the last fragment uses a different font/style than the current setting sometimes increases the size, sometimes just moves the q to a different location, but often also reduces it because the following text doesn't need to set the font.
* In some cases, Tw was added seperately to the rest of the content. Fixing that moved it to a different location, but usually didn't change the length (sometimes it elminated a newline).
* Found a way to save 3 bytes per word with unicode fonts (merging the leading space with the following word).

Since there were worries about increasing the file size of the resulting PDFs, here's a rundown of the affected test files:

same size:
* moving Tw - render_styled_newpos.pdf
* moving Tw - multi_cell_newpos.pdf
* moving Tw - multi_cell_ln_newpos.pdf
* moving Tw - multi_cell_markdown_with_fill_color.pdf
* moving Tw - multi_cell_markdown_justified.pdf
* moving Tw - simple_outline.pdf
* moving qQ - graphics_context.pdf
* moving qQ - links.pdf
* moving qQ - link_with_zoom_and_shift.pdf
* moving qQ - html_features.pdf
* moving qQ - html_justify_paragraph.pdf
* moving qQ - flextemplate_rotation.pdf
* moving qQ - flextemplate_multipage.pdf
* moving qQ - flextemplate_elements.pdf
* moving qQ - template_nominal_hardcoded.pdf
* moving qQ - template_nominal_csv.pdf
* moving qQ - template_nominal_multipage.pdf
* moving qQ - template_nominal_justify.pdf

shorter
* moving Tw - 2_pages_outline.pdf
* moving Tw - ln_positioning_and_page_breaking_for_multicell.pdf
* replace F with qQ - cell_markdown_bleeding.pdf
* replace F with qQ - cell_markdown_right_aligned.pdf
* ttf words - multi_cell_j_paragraphs
* moving qQ - test_img_inside_html_table_centered_with_align.pdf

longer
* add qQ - multi_cell_markdown
* add qQ /ttf words - multi_cell_justified_with_unicode_fonts
* add qQ /ttf words - multi_cell_markdown_with_ttf_fonts

### CODE REVIEW

**Comment by Lucas-C** [comment]
This PR includes 88 commits:
you probably should rebase your branch on the `master` branch of this repo

**Comment by gmischler** [comment]
> This PR includes 88 commits: you probably should rebase your branch on the `master` branch of this repo

I've tried, but Git Desktop repeatedly got confused and I had to cancel the attempt...

**Comment by Lucas-C** [comment]
I performed the rebase. Could you check that all is right @gmischler please?

**Comment by gmischler** [comment]
> I performed the rebase. Could you check that all is right @gmischler please?
> 
> Also, seems like many reference PDF files have been impacted by this change: could you explain how / why?

I was just doing another rebase attempt myself again, which may have conflicted with yours...

There now seem to be a few files included that shouldn't be.
I'll have to comb through everything to make sure it's actually in a useable state.

And PDFs that aren't listed and explained at the top here should probably not be included.

**Comment by gmischler** [comment]
Ok, I think it's clean now.

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -2506,8 +2489,17 @@ def _render_styled_cell_text(
...
+            # If any PDF settings have been left modified, wrap the line in a local context.
+            if (
+                current_font_style != self.font_style
+                or self.fill_color != self.text_color
+            ):
+                s = f"q {' '.join(sl)} Q"
+                self._out(s)
+            else:
+                s = " ".join(sl)
+                self._out(s)
```
Comment (Lucas-C) [inline fpdf/fpdf.py:2502]: This line is shared between the 2 parts of the `if` and could be de-indented

File: fpdf/fpdf.py [hunk 2]
```diff
@@ -2406,98 +2406,81 @@ def _render_styled_cell_text(
...
                     if frag.underline:
                         underlines.append((self.x + dx + s_width, frag.string))
-                    self.underline = frag.underline
-                    s_width += self.get_string_width(
-                        frag.string, True
-                    ) + self.ws * frag.string.count(" ")
-            s += " ET"
-            # Restoring font style & underline mode after handling changes
-            # by Markdown annotations:
-            if not self._markdown_leak_end_style:
```
Comment (Lucas-C) [inline fpdf/fpdf.py:2483]: Seems that you got rid of `self._markdown_leak_end_style`: great, one less hacky attribute!
Could you please get rid of all references to this?

**Review by Lucas-C — COMMENTED** [review]
The code seems quite clearer after this PR, nice and good job!

Thank you very much for the rundown on the affected test files sizes.

However it does not seem totally accurate...
`render_styled_newpos.pdf` for example is 15bytes larger.
Would you know why?

File: fpdf/fpdf.py [hunk 3]
```diff
@@ -2406,98 +2406,81 @@ def _render_styled_cell_text(
...
 
                     # Determine the position of space (" ") in the current subset and
                     # split words whenever this mapping code is found
                     words = txt_frag_mapped.split(
-                        chr(self.current_font["subset"].pick(ord(" ")))
+                        chr(current_font["subset"].pick(ord(" ")))
                     )
 
-                    s += " ["
+                    wsl = []
```
[the remark this answers is not in the document]
Comment (gmischler) [reply fpdf/fpdf.py:2427]: done.

File: fpdf/fpdf.py [hunk 4]
```diff
@@ -2406,98 +2406,81 @@ def _render_styled_cell_text(
...
+                            wsl.append(f"{adj:.3f}({space}{word})")
+                    sl.append(f"[{' '.join(wsl)}] TJ")
                     if frag.underline:
                         underlines.append((self.x + dx + s_width, frag.string))
-                    self.underline = frag.underline
-                    s_width += self.get_string_width(
-                        frag.string, True
-                    ) + word_spacing * frag.string.count(" ")
-                    s += "] TJ"
+                    sw = self.get_normalized_string_width_with_style(
```
[the remark this answers is not in the document]
Comment (gmischler) [reply fpdf/fpdf.py:2438]: done.

File: fpdf/fpdf.py [hunk 5]
```diff
@@ -2406,98 +2406,81 @@ def _render_styled_cell_text(
...
+                        frag.string, current_font_style
+                    )
+                    sw *= self.font_stretching * self.font_size * 0.00001  # /100 /1000
```
[the remark this answers is not in the document]
Comment (gmischler) [reply fpdf/fpdf.py:2441]: done.

File: test/text/test_multi_cell.py [hunk 6]
```diff
@@ -284,3 +285,23 @@ def test_multi_cell_j_paragraphs(tmp_path):  # issue 364
 
     pdf.multi_cell(w=0, h=None, txt=text, align="J")
     assert_pdf_equal(pdf, HERE / "multi_cell_j_paragraphs.pdf", tmp_path)
+
+
+def test_multi_cell_font_leakage(tmp_path):  # Issue #359
```
[the remark this answers is not in the document]
Comment (gmischler) [reply test/text/test_multi_cell.py:290]: Huh, thought I had done that. Now for real.

File: fpdf/fpdf.py (hunk 1 above)
Comment (gmischler) [reply fpdf/fpdf.py:2502]: yup.

File: fpdf/fpdf.py (hunk 2 above)
Comment (gmischler) [reply fpdf/fpdf.py:2483]: Huh, thought I had done that. Now for real.

**Comment by gmischler** [comment]
> Thank you very much for the rundown on the affected test files sizes.
> 
> However it does not seem totally accurate... `render_styled_newpos.pdf` for example is 15bytes larger. Would you know why?

Interesting.
I had compared the uncompressed files, which are the same length. Apparently in this case the compression is slightly less effective for the new version. I would expect such an effect to average out in total, though.

**Comment by Lucas-C** [comment]
I'm OK to merge this one the conflicts are resolved

**Comment by Lucas-C** [comment]
Merged: thank you @gmischler !

**Pull request merged**